rfd/0150-ssh-connection-resumption.md

```diff
@@ -0,0 +1,127 @@
+
+### State
+
+Both sides need to maintain a log of data that has been generated by the application and potentially sent over the wire, but hasn't yet been acknowledged by the peer. Such a buffer should be limited in size to avoid unconstrained resource consumption - from some early tests, a buffer size of 2MiB seems to be sufficient to maintain near full bandwidth when transferring data across a single SSH channel, compared to a regular SSH connection that doesn't make use of the resumption protocol. The protocol does not mandate the use of a buffer with a fixed maximum size, so other approaches can be used in the future if need be.
+
+This design allows for keepalive frames containing no acknowledgement and no data, similarly to the protocol used in [TLS routing Ping](https://github.com/gravitational/teleport/blob/master/rfd/0081-tls-ping.md); a frame consisting of two NULs should be sent if no data has been otherwise sent for some _keepalive interval_ (30 seconds?), and the underlying connection should be closed and considered dead if no data has been received for two or three times that interval.
+
+The server side of the connection is tasked with keeping track of the existing resumable connections, and should take care not to let dead connections accumulate if they're unused. A connection that has no underlying transport for longer than a _grace period_ - we have an idle timeout of 15 minutes for SSH connections, but we can pick something slightly shorter, around 5 minutes - should be closed and then cleaned up by the server. Closed connections (as a result of a SSH error, or as a result of the peer sending a close signal) should likewise be cleaned up.
+
+The client side of the connection is tasked with reconnecting to the server; it should do so whenever the underlying connection has a fault, and it should also do so periodically, so that it's possible to rotate the Teleport proxy services forwarding the connection gracefully and without the connection being impacted. A _reconnection interval_ of 3 minutes works well for Teleport Cloud clusters, as the Teleport proxy is forcibly terminated 5 minutes after beginning its shutdown (which happens after new proxies are already available for connection routing). Similar operational recommendations should be documented for on-prem Teleport clusters.
```

Comment:
Per-user limits, from what I can tell, are only enforced at the SSH connection level (after a handshake), so those are not going to affect connection resumption, and they won't be affected other than the fact that it's quite a bit easier to accidentally leave a session open and hanging: no matter what we implement, if a user opens a connection with `tsh`, then disconnects from the internet, then kills `tsh`, then they won't have any way to join that same connection, and if the connection limits are very low, the user might just end up locked out for 5 or 15 minutes depending on which timeout we pick.

In those scenarios I think that the only thing we can do is provide a `--no-resume` option so that `tsh` will open a regular, non-resumable connection that will hopefully die within a couple of minutes upon encountering a network problem.